cpp/include/raft/core/detail/copy.hpp

```diff
@@ -458,6 +471,7 @@ mdspan_copyable_t<DstType, SrcType> copy(resources const& res, DstType&& dst, Sr
 #endif
   } else if constexpr (config::custom_kernel_allowed) {
 #ifdef __CUDACC__
+    config::check_for_unique_dst(dst);
```

Comment:
@mhoemmen How do we feel about this for the uniqueness check? It should only be necessary in this branch, since it's the only parallelized copy that is not already linked exclusively to layout_left/layout_right mdspans.

Reply:
@wphicks Thanks for asking! : - )  It might be correct given the current state of the code, but I could see future changes breaking this.  For example, some developer might later want to add support for strided layouts (i.e., layouts for which `is_strided()` is `true`) other than `layout_stride`.  `layout_stride` has preconditions that no stride can be zero, but strided layouts in general do not have this precondition.  This means that a strided layout could be nonunique (because it has a zero stride, but a nonzero extent). 

In my view, it would be safer to move this precondition check to the top of the `copy` function.  `is_unique()` is meant to be cheap (that's why it's conservative), and the check is compiled out by the `if constexpr (! m.is_always_unique())` call.  It also makes the code more self-documenting to check all the preconditions at the top of the function.

Reply:
My thought in narrowing the scope of the check was to avoid artificially limiting the cases in which the copy can fail. I agree that the performance impact won't be significant.

Right now the code is structured around determining what method can be used to perform the copy. If we determine that a particular non-unique layout can be safely copied to with e.g. std::copy, I'd rather not prevent that up front. A fine example of this would be copying underlying memory of between identical non-unique layouts. That's safe even if the destination is non-unique.

If a user copies an mdspan from a layout which contains unique values for destination locations which are non-unique, I'd say that's user error and fine to leave as undefined behavior (though we should probably explicitly call that out in documentation).

Our remaining responsibility, then is to ensure that well-defined copies to non-unique destinations are correct. For parallel copies, that is not a simple task, so I think it makes sense to just fail for now. The only two parallel copy methods we have are cuBlas and the custom kernel. cuBlas can inherently only be invoked for unique layouts, so we are left with failing on the custom kernel. If a future developer adds additional parallel techniques, it's up to them to determine whether or not they can handle well-defined copies to non-unique layouts.

Reply:
Just added a warning about ill-defined copies to non-unique layouts to the docs.

Reply:
I'll just warn that most developers may not remember that non-unique layouts can exist : - ) .